articles/changes.md

```diff
@@ -0,0 +1,257 @@
+In ROS 2 a Python package can use anything in setup.py files, e.g. entry points since they are being invoked with `python3 setup.py install`.
+
+
+### Environment setup
+
+In ROS 1 the build tool generates scripts which must be sourced in order to setup the environment before being able to use the built ROS packages.
+This approach only works when the ROS packages are being built with ROS specific build tool.
+
+In ROS 2 the environment setup is separated into package-specific scripts and workspace-specific scripts.
+Each package provides the necessary scripts to make itself usable after being built.
+The build tool only invokes the workspace-specific scripts which then call the package-specific scripts.
+
+
+### No non-isolated build
+
+In ROS 1 multiple packages can be built in a single CMake context.
+While this speeds up the build step, every package needs to ensure that cross package target dependencies are defined correctly.
+Additionally all packages share the same namespace which leads to colliding target names, etc.
+
+In ROS 2 only isolated builds are supported, i.e. every package is built separately on its own.
+
+
+### No devel space
+
+In ROS 1 packages can be built without installing them.
+From the devel space in combination with the source space the system is already usable.
+But every package has to actively support the devel space, e.g. in environment hooks and CMake code.
+
+In ROS 2 a package must be installed after building it before it can be used.
+
+One reason for the devel space in ROS 1 is to enable the developer to change files, e.g. Python code or launch files, and use the modified code directly without the need to rebuild the package.
+This benefit is preserved in ROS 2 by optionally replacing copy operations in the install steps with symlinks.
+
+
+### Support catkin_simple use case
+
+In ROS 1 the package catkin_simple is aiming to make writing the CMake code of ROS packages easier.
+In many cases it is not achieving this goal which is often due to restrictions of the design necessary for support features like the devel space.
+
+In ROS 2 the CMake API was restructured to support this use case.
+
+
+### Minimal support for packages without manifest
+
+In ROS 1 only packages with a manifest file are considered by the build system.
+In ROS 2 it is possible to detect packages with supported build system in folders without a manifest file.
+If the package follows common practice it might even be possible to detect some of the missing meta information (like dependencies).
+
+
+## Messages, Services
+
+For more information please see the [ROS interface definition](http://design.ros2.org/articles/interface_definition.html) article.
+
+
+### Separated namespaces
+
+In ROS 1 .msg and .srv files can have the same name but the generated code collides.
+The same is the case for the request and response parts of services.
+
+In ROS 2 the generated code uses separate namespaces to guarantee it is collision-free.
```

Comment:
> They introduce gratuitous differences between ROS 1 and ROS 2 in order to solve a merely theoretical "problem".

I don't think it is a theoretical problem, since we were forced to create the `std_srvs` package just for this case in ROS 1, i.e. there is an `std_msgs/Empty.msg` as well as an `std_srvs/Empty.srv`. Both of which are valid in my opinion and being forced to put them in separate packages is an artificial constraint from my perspective.

> The current ROS 1 <-> ROS 2 bridge is a pretty practical example though where this is necessary.

To be fair this is a side effect of this change, not really a rational, but it is a convenient solution. The goal of having different include paths and namespaces for messages for the bridge could be solved some other way, just not nearly as elegantly.

----

I was thinking about a compromise for this issue and I think I have a proposal.

In ROS 2 we could generate a file for each service and messages that is like the ROS 1 include, e.g. `#include <std_msgs/String.h>`, the contents of which would look something like this:

```c++
#include <std_msgs/msg/string.hpp>

namespace std_msgs
{
using std_msgs::msg::String;
using std_msgs::msg::String_;
// ... and other symbols in message headers.
}
```

That way you can continue to include messages in ROS 2 as you did in ROS 1 and you get the same symbols (mostly). For the case of `Empty.msg` and `Empty.srv` the `<std_msgs/Empty.h>` could contain this:

```c++
#error Including <std_msgs/Empty.h> is ambiguous since there is a Message and a Service by that name, use either <std_msgs/msg/empty.hpp> or <std_msgs/srv/empty.hpp>.
```

Some points about the approach:

- We can allow opting in/out of this header by installing them to a place like `<prefix>/include/ros1_compat/std_msgs/Empty.h` and conditionally adding `<prefix>/ros1_compat` to the include directories.
- The bridge can explicitly opt out of this, allowing it to have different include paths for ROS 1 and ROS 2.
- We can choose whether or not to put a deprecation message in the compat header to encourage people to migrate to the new format, which I still believe is better because it is more conceptually consistent and provides more information to people reading unfamiliar code.
- We can provide similar symbols to what is in ROS 1, but somethings will likely end up being different, so I'm still not optimistic that we can achieve a zero change migration w.r.t. messages.